service/history/workflow_task_handler_callbacks.go

```diff
@@ -884,8 +904,7 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 	}
 
 	resp := &historyservice.RespondWorkflowTaskCompletedResponse{}
-	//nolint:staticcheck
-	if request.GetReturnNewWorkflowTask() && newWorkflowTask != nil {
+	if newWorkflowTask != nil && bypassTaskGeneration {
```

Comment:
I don't know about this change... what if newWorkflowTask != nil but bypassTaskGeneration is false? then shouldn't we send the workflow task? oh, I think I see the point here... it's that newWorkflowTask could refer to a task that's only scheduled, or one that's scheduled and started. and we shouldn't send it if only started.

that's fine, but what about the case on line 878 where we create a speculative task? is it ever possible that bypassTaskGeneration is false in that case?

maybe it should be 
```suggestion
	if newWorkflowTask != nil && (bypassTaskGeneration || newWorkflowTaskType == enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE) {
```
?

or maybe it would be better to have another bool that set to true if we add a started event (either normal or speculative)?

Reply:
I admit bypassTaskGeneration is confusing, by this method is so long and complex anyways. In short, I don't think with this code it's possible to have a speculative task with bypassTaskGeneration=false.
I'll try to separate the bools if I find time.

Reply:
this is definitely the worst+scariest function in the codebase :slightly_frowning_face: 

should we rename bypassTaskGeneration to returnImmediateWorkflowTask or something? and yeah there are some other ways this could be improved (later).

I think I agree that because of line 718-721 we can't have speculative and bypassTaskGeneration==false. line 859 is confusing in that case since it uses bypassTaskGeneration as it could be either value.. it might be clearer to just pass `true` with a comment saying bypassTaskGeneration must be true there.